File: app/src/main/java/com/relic/presentation/adapter/PostItemAdapter.java
package com.relic.presentation.adapter;

import android.databinding.BindingAdapter;
import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.relic.R;
import com.relic.data.models.PostModel;
import com.relic.databinding.PostItemBinding;
import com.squareup.picasso.Picasso;

import java.util.List;

public class PostItemAdapter extends RecyclerView.Adapter<PostItemAdapter.PostItemVH>{
  private final String TAG = "POST_ADAPTER";
  private List<PostModel> postList;
  private PostItemOnclick onClick;
  private ImageOnClick onClickImage;


  /**
   * Viewholder item for storing post view bindings
   */
  class PostItemVH extends RecyclerView.ViewHolder {
    // since databinding, the viewholder only needs to contain autogenerated binding object
    final PostItemBinding postBinding;

    public PostItemVH(PostItemBinding postBinding) {
      super(postBinding.getRoot());
      this.postBinding = postBinding;
    }
  }

  public PostItemAdapter(PostItemOnclick onClick, ImageOnClick onClickImage) {
    super();
    this.onClick = onClick;
    this.onClickImage = onClickImage;
  }

  @NonNull
  @Override
  public PostItemVH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    // Extra declaration for clarity and readability
    PostItemBinding postBinding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()),
        R.layout.post_item, parent, false);

    // attach the onClick s to the binding
    postBinding.setPostOnClick(this.onClick);
    postBinding.setPostOnClickImage(this.onClickImage);
    return new PostItemVH(postBinding);
  }

  @Override
  public void onBindViewHolder(@NonNull PostItemVH holder, int position) {
    // attach the current post to the viewholder's post binding
    holder.postBinding.setPostModel(postList.get(position));
    holder.postBinding.executePendingBindings();
  }


  @Override
  public int getItemCount() {
    return postList == null ? 0 : postList.size();
  }


  public void setPostList(List<PostModel> newPostList) {
    if (postList == null) {
      postList = newPostList;
      notifyItemRangeRemoved(0, newPostList.size());
    } else {
      // used to tell list what has changed
      DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new DiffUtil.Callback() {
        @Override
        public int getOldListSize() {
          return postList.size();
        }

        @Override
        public int getNewListSize() {
          return newPostList.size();
        }

        @Override
        public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
          return postList.get(oldItemPosition).getId().equals(newPostList.get(newItemPosition).getId());
        }

        @Override
        public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
          return postList.get(oldItemPosition).getId().equals(newPostList.get(newItemPosition).getId());
        }
      });
      // sets the new list as the current one
      postList = newPostList;
      diffResult.dispatchUpdatesTo(this);
    }
  }


  /**
   * Clears all the data from the current list and tells list to show as empty
   */
  public void resetPostList() {
    if (postList != null) {
      postList.clear();
      notifyDataSetChanged();
    }
  }


  @BindingAdapter({"bind:thumbnail"})
  public static void loadThumbnail(ImageView imgView, String thumbnailUrl) {
    if (thumbnailUrl != null && thumbnailUrl.length() != 0) {
      // does not load image if the banner img string is empty
      try {
        Log.d("POSTITEM_ADAPTER", "URL = " + thumbnailUrl);
        Picasso.get().load(thumbnailUrl).fit().centerCrop().into(imgView);
      }
      catch (Error e) {
        Log.d("POSTITEM_ADAPTER", "Issue loading image " + e.toString());
      }
    }
  }

}

File: app/src/main/java/com/relic/presentation/displaysub/list/PostItemAdapter.kt
package com.relic.presentation.displaysub.list

import android.support.v7.util.DiffUtil
import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import com.relic.data.models.PostModel
import com.relic.presentation.customview.RelicPostItemView
import com.relic.presentation.displaysub.DisplaySubContract

class PostItemAdapter (
        private val postAdapterDelegate : DisplaySubContract.PostAdapterDelegate
) : RecyclerView.Adapter <PostItemVH> (), DisplaySubContract.PostItemAdapterDelegate {
    private var postList: List<PostModel> = ArrayList()

    override fun getItemCount() = postList.size

    override fun onCreateViewHolder(parent: ViewGroup, position: Int): PostItemVH {
        return PostItemVH(RelicPostItemView(parent.context)).apply {
            initializeOnClicks(this@PostItemAdapter)
        }
    }

    override fun onBindViewHolder(viewholder: PostItemVH, position: Int) {
        viewholder.bindPost(postList[position], position)
    }

    fun clear() { setPostList(emptyList()) }

    fun setPostList(newPostList: List<PostModel>) {
        DiffUtil.calculateDiff(object : DiffUtil.Callback() {
            override fun getOldListSize(): Int {
                return postList.size
            }

            override fun getNewListSize(): Int {
                return newPostList.size
            }

            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                return postList[oldItemPosition].id == newPostList[newItemPosition].id
            }

            override fun areContentsTheSame(
                    oldItemPosition: Int,
                    newItemPosition: Int
            ): Boolean {
                val oldPost = postList[oldItemPosition]
                val newPost = newPostList[newItemPosition]

                return oldPost.id == newPost.id && oldPost.isVisited == newPost.isVisited
            }
        }).dispatchUpdatesTo(this)

        postList = newPostList
    }

    // start region for onclick handlers

    override fun onPostPressed (itemPosition : Int) {
        postList[itemPosition].also {
            // update the view and local model to reflect onclick
            it.isVisited = true
            // update post to show that it has been visited
            postAdapterDelegate.visitPost(it.id, it.subreddit)
        }
        notifyItemChanged(itemPosition)
    }

    // initialize onclick for the upvote button
    override fun onPostUpvotePressed(itemPosition : Int) {
        postList[itemPosition].also {
            // determine the new vote value based on the current one and change the vote accordingly
            val newStatus = if (it.userUpvoted <= 0) 1 else 0

            // optimistic, update copy cached in adapter and make request to api to update in server
            it.userUpvoted = newStatus
            postAdapterDelegate.voteOnPost(it.id, newStatus)
        }
        notifyItemChanged(itemPosition)
    }

    // initialize onclick for the downvote button
    override fun onPostDownvotePressed(itemPosition : Int) {
        postList[itemPosition].also {
            // determine the new vote value based on the current one and change the vote accordingly
            val newStatus = if (it.userUpvoted >= 0) -1 else 0

            // optimistic, update copy cached in adapter and make request to api to update in server
            it.userUpvoted = newStatus
            postAdapterDelegate.voteOnPost(it.id, newStatus)
        }
        notifyItemChanged(itemPosition)
    }

    override fun onPostSavePressed (itemPosition : Int) {
        postList[itemPosition].also {
            // calculate new save value based on the previous one and tell vm to update appropriately
            val newStatus = !it.saved
            postAdapterDelegate.savePost(it.id, newStatus)

            // update the view and local model to reflect onclick
            it.saved = newStatus
        }
        notifyItemChanged(itemPosition)
    }

    override fun onPostLinkPressed (itemPosition : Int) {
        postAdapterDelegate.onThumbnailClicked(postList[itemPosition].url)
    }

    // end region for onclick handlers
}

